_posts/2024/2024-02-06-about-caching-in-eo.md

```diff
@@ -0,0 +1,263 @@
+
+### ccache/sccache
+In compiled programming languages, building a project containing many source code files takes a long time.
+This time is spent on loading of libraries, preparing, optimizing, checking the code, and so on.
+To speed up the assembly of compiled languages, [ccache](https://ccache.dev) 
+and [sccache](https://github.com/mozilla/sccache) are used.
+Let's look at the assembly scheme using C++ as an example
+to imagine the build process in compiled languages:
+
+<p align="center">
+  <img src="/images/defaultCPhase.svg">
+</p>
+
+1) First, preprocessor gets the input files. The input files are source files `.cpp` and header files `.h`.
+The result is a single edited file `.cpp` with human-readable code that the compiler will get.
+
+
+2) The compiler receives the edited code file `.cpp` and converts it into machine code, presented in an object file.
+At the compilation stage, parsing occurs, which checks whether the code matches
+rules of a specific programming language.
+At the end, the compiler optimizes the resulting machine code and produces an object file. 
+To speed up compilation, different files of the same project are compiled in parallel.
+
+3)  Then, the [Linker](https://en.wikipedia.org/wiki/Linker_(computing)) gets object files.
+The result of the linker is an executable `.exe` file.
+
+
+As a result, in compiled languages, multiple files are simultaneously and independently converted
+into machine code at the compilation stage.
+This machine code is then combined into one executable file.
+
+
+`ccache` hash algorithm, for the hashing of information to find cached files fast. 
+The [`ccache` hash](https://ccache.dev/manual/4.8.2.html#_common_hashed_information)
+includes information:
+* the file contents
+* the current directory of the file
+* the name of the compiler
+* the compiler’s size and modification time
+* extensions used by the compiler. 
+
+
+A compressed machine code file is placed in the cache using the received key.
+
+
+`ccache` has two main caching methods:
+1) `Direct mode` - hash is generated based on the source code. 
+`Direct mode` compiles the program faster, since the preprocessor step is skipped.
+However,the header files are not checked for changes, so the project may be built with not verified header files.
+2) `Preprocessor mode` - hash is generated based on the result of preprocessor.
+`Preprocessor mode` is slower than `direct mode`, but the project is built with verified header files.
+
+`Sccache`, unlike `ccache`, allows you to store cached files not only locally, but also in a cloud data storage.
+And `sccache` includes support for caching the compilation of C/C++ code, 
+[Rust](https://github.com/mozilla/sccache/blob/main/docs/Rust.md), as well as NVIDIA's CUDA using 
+[nvcc](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html),
+and [clang](https://llvm.org/docs/CompileCudaWithLLVM.html), while `ccache` works with C and C++ code.
```

Comment:
@Yanich96 What do you mean:

> `ccache` works with C and C++ code.

As I undestand `ccache` works with compilers, not with code.